system-parachains/coretime/coretime-kusama/src/coretime.rs

```diff
@@ -0,0 +1,287 @@
+				call: credit_account_call.encode().into(),
+			},
+		]);
+
+		match PolkadotXcm::send_xcm(Here, Location::parent(), message) {
+			Ok(_) => log::debug!(
+				target: "runtime::coretime",
+				"Instruction to credit account sent successfully."
+			),
+			Err(e) => log::error!(
+				target: "runtime::coretime",
+				"Instruction to credit account failed to send: {:?}",
+				e
+			),
+		}
+	}
+
+	fn assign_core(
+		core: CoreIndex,
+		begin: RCBlockNumberOf<Self>,
+		assignment: Vec<(CoreAssignment, PartsOf57600)>,
+		end_hint: Option<RCBlockNumberOf<Self>>,
+	) {
+		use crate::coretime::CoretimeProviderCalls::AssignCore;
+		let assign_core_call =
+			RelayRuntimePallets::Coretime(AssignCore(core, begin, assignment, end_hint));
+
+		// Weight for `assign_core` from Kusama runtime benchmarks:
+		// `ref_time` = 10177115 + (1 * 25000000) + (2 * 100000000) + (80 * 13932) = 236291675
+		// `proof_size` = 3612
+		// Add 5% to each component and round to 2 significant figures.
+		let call_weight = Weight::from_parts(248_000_000, 3800);
+
+		let message = Xcm(vec![
+			Instruction::UnpaidExecution {
+				weight_limit: WeightLimit::Unlimited,
+				check_origin: None,
+			},
+			Instruction::Transact {
+				origin_kind: OriginKind::Native,
+				require_weight_at_most: call_weight,
+				call: assign_core_call.encode().into(),
+			},
+		]);
+
+		match PolkadotXcm::send_xcm(Here, Location::parent(), message) {
+			Ok(_) => log::debug!(
+				target: "runtime::coretime",
+				"Core assignment sent successfully."
+			),
+			Err(e) => log::error!(
+				target: "runtime::coretime",
+				"Core assignment failed to send: {:?}",
+				e
+			),
+		}
+	}
+
+	fn check_notify_revenue_info() -> Option<(RCBlockNumberOf<Self>, Self::Balance)> {
+		let revenue = CoretimeRevenue::get();
+		CoretimeRevenue::set(&None);
+		revenue
+	}
+
+	#[cfg(feature = "runtime-benchmarks")]
+	fn ensure_notify_revenue_info(when: RCBlockNumberOf<Self>, revenue: Self::Balance) {
+		CoretimeRevenue::set(&Some((when, revenue)));
+	}
+}
+
+/// Implements the [`AdaptPrice`] trait to control the price changes of bulk coretime. This tweaks
+/// the [`pallet_broker::Linear`] implementation which hard-corrects to 0 if cores are offered but
+/// not sold for just one sale. The monotonic lead-in is increased in magnitude to enable faster
+/// price finding. The change in base price between sales has a lower limit of 0.5 to allow downward
+/// pressure to be applied, while keeping a conservative upper limit of 1.2 (movements capped at 20%
+/// if cores sell out) to avoid runaway prices in the early sales. The intention is that this will
+/// be coupled with a low number of cores per sale and a 100% ideal bulk ratio for the first sales.
+pub struct LinearPlusC;
+impl AdaptPrice for LinearPlusC {
+	fn leadin_factor_at(when: FixedU64) -> FixedU64 {
+		// Start at 5x the base price and decrease to the base price at the end of leadin.
+		FixedU64::from(5).saturating_sub(when)
```

Comment:
I think this should be `saturating_sub(4*when)`

Reply:
I think we should have tests here.

Reply:
🤦‍♂️

Reply:
Fixed and added tests